File: app/src/main/java/com/flynes/emu/gamecenter/GameCenterState.java
package com.flynes.emu.gamecenter;

import com.flynes.emu.Popularity;

import java.text.Normalizer;
import java.util.ArrayList;
import java.util.Collections;
import java.util.EnumMap;
import java.util.List;
import java.util.Locale;
import java.util.HashSet;

/** Pure, deterministic navigation state for the landscape Game Center. */
public final class GameCenterState {
    public enum Category { RECENT, FAVORITES, ALL, BUILTIN }

    /**
     * Shared, versioned two-player capability projection (design 2026-09-13 §3.2).
     * The only source of two-player eligibility: platforms never infer it from
     * filenames, "P2" in a title, or controller counts. Unread ids and entries
     * whose profile version mismatches project UNKNOWN, never UNSUPPORTED.
     */
    public enum MultiplayerEligibility { UNSUPPORTED, SUPPORTED, UNKNOWN }

    public static final class MultiplayerCapabilityRegistry {
        private static final class Entry {
            final MultiplayerEligibility eligibility;
            final long profileVersion;
            Entry(MultiplayerEligibility eligibility, long profileVersion) {
                this.eligibility = eligibility;
                this.profileVersion = profileVersion;
            }
        }

        private final long profileVersion;
        private final java.util.HashMap<String, Entry> entries = new java.util.HashMap<>();

        public MultiplayerCapabilityRegistry(long profileVersion) {
            this.profileVersion = profileVersion;
        }

        public void put(String canonicalId, MultiplayerEligibility eligibility, long profileVersion) {
            entries.put(canonicalId, new Entry(eligibility, profileVersion));
        }

        public MultiplayerEligibility eligibilityFor(String canonicalId) {
            Entry entry = entries.get(canonicalId);
            if (entry == null || entry.profileVersion != profileVersion) return MultiplayerEligibility.UNKNOWN;
            return entry.eligibility;
        }
    }

    private Category category = Category.ALL;
    private String query = "";
    private boolean multiplayerOnly = false;
    private final EnumMap<Category, String> selections = new EnumMap<>(Category.class);
    private List<GameCenterItem> rankedInput = Collections.emptyList();
    private List<GameCenterItem> rankedItems = Collections.emptyList();
    public GameCenterState() { }

    public static GameCenterState restore(
            String categoryName, String query, String selectedCanonicalId) {
        return restore(categoryName, query, selectedCanonicalId, false);
    }

    public static GameCenterState restore(
            String categoryName, String query, String selectedCanonicalId, boolean multiplayerOnly) {
        GameCenterState state = new GameCenterState();
        try {
            state.category = Category.valueOf(categoryName);
        } catch (RuntimeException ignored) {
            state.category = Category.ALL;
        }
        state.query = query == null ? "" : query;
        state.multiplayerOnly = multiplayerOnly;
        if (selectedCanonicalId != null && !selectedCanonicalId.trim().isEmpty()) {
            state.selections.put(state.category, selectedCanonicalId);
        }
        return state;
    }

    public Category category() { return category; }
    public String query() { return query; }
    public String selectedCanonicalId() { return selections.get(category); }

    /** Independent two-player filter: persisted per device, never changed by
     *  category, query, or connection events (design U04). */
    public void setMultiplayerOnly(boolean value) { multiplayerOnly = value; }
    public boolean multiplayerOnly() { return multiplayerOnly; }

    public void setCategory(Category value) {
        category = value == null ? Category.ALL : value;
    }

    public void setQuery(String value) {
        query = value == null ? "" : value;
    }

    public void select(String canonicalId) {
        if (canonicalId == null || canonicalId.trim().isEmpty()) selections.remove(category);
        else selections.put(category, canonicalId);
    }

    public List<GameCenterItem> itemsFor(Category value, List<GameCenterItem> all) {
        // Copy the input when caching: callers may replace rows in a mutable list in place.
        if (value == Category.ALL && rankedInput.equals(all)) return rankedItems;
        ArrayList<GameCenterItem> result = new ArrayList<>();
        HashSet<String> seen = new HashSet<>();
        for (GameCenterItem item : all) {
            boolean include = switch (value) {
                case RECENT -> item.lastPlayedSequence() > 0;
                case FAVORITES -> item.favorite();
                case ALL -> true;
                case BUILTIN -> item.builtin();
            };
            // Canonical IDs derive from ROM payload hashes, never from title similarity.
            if (include && seen.add(item.canonicalId())) result.add(item);
        }
        if (value == Category.RECENT) {
            result.sort((left, right) -> Long.compare(
                    right.lastPlayedSequence(), left.lastPlayedSequence()));
        } else if (value == Category.ALL) {
            java.util.HashMap<String, Integer> scores = new java.util.HashMap<>();
            for (GameCenterItem item : all) {
                int score = item.popularityScore() >= 0 ? item.popularityScore()
                        : Math.max(Popularity.score(item.titleEn()), Math.max(
                        Popularity.score(item.titleZhHans()), Popularity.score(item.originalFilename())));
                scores.merge(item.canonicalId(), score, Math::max);
            }
            result.sort((left, right) -> {
                int order = Integer.compare(scores.get(right.canonicalId()), scores.get(left.canonicalId()));
                // Content IDs are stable ASCII hashes: identical tie order on all platforms.
                return order != 0 ? order : left.canonicalId().compareTo(right.canonicalId());
            });
        }
        List<GameCenterItem> immutable = Collections.unmodifiableList(result);
        if (value == Category.ALL) {
            rankedInput = new ArrayList<>(all);
            rankedItems = immutable;
        }
        return immutable;
    }

    public List<GameCenterItem> filtered(List<GameCenterItem> all) {
        List<GameCenterItem> categoryItems = itemsFor(category, all);
        String needle = normalize(query);
        if (needle.isEmpty()) return categoryItems;
        ArrayList<GameCenterItem> result = new ArrayList<>();
        for (GameCenterItem item : categoryItems) {
            if (contains(item.titleEn(), needle) || contains(item.titleZhHans(), needle)
                    || contains(item.originalFilename(), needle)) result.add(item);
        }
        return Collections.unmodifiableList(result);
    }

    /** Applies the two-player filter AFTER the existing category/search filter
     *  and sort, preserving the relative order of the surviving items. UNKNOWN
     *  and UNSUPPORTED games drop out only while the filter is on. */
    public List<GameCenterItem> filtered(List<GameCenterItem> all, MultiplayerCapabilityRegistry registry) {
        List<GameCenterItem> base = filtered(all);
        if (!multiplayerOnly) return base;
        ArrayList<GameCenterItem> result = new ArrayList<>();
        for (GameCenterItem item : base) {
            if (registry.eligibilityFor(item.canonicalId()) == MultiplayerEligibility.SUPPORTED) {
                result.add(item);
            }
        }
        return Collections.unmodifiableList(result);
    }

    public void reconcile(List<GameCenterItem> visible) {
        String selected = selectedCanonicalId();
        boolean exists = false;
        for (GameCenterItem item : visible) exists |= item.canonicalId().equals(selected);
        if (!exists) select(visible.isEmpty() ? null : visible.get(0).canonicalId());
    }

    private static boolean contains(String value, String needle) {
        return normalize(value).contains(needle);
    }

    private static String normalize(String value) {
        return Normalizer.normalize(value == null ? "" : value, Normalizer.Form.NFKC)
                .toLowerCase(Locale.ROOT).trim();
    }
}

File: shared/src/product/game_center_state.cpp
#include "flynes/product/game_center_state.hpp"

#include <algorithm>
#include <cctype>
#include <string_view>
#include <unordered_set>
#include <unordered_map>

namespace flynes::product {
namespace {

struct PopularityRank {
    int score;
    std::vector<std::string_view> keywords;
};

const PopularityRank popularity_ranks[] = {
#define FLYNES_POPULARITY_RANK(score, ...) {score, {__VA_ARGS__}},
#include "../../data/popularity.inc"
#undef FLYNES_POPULARITY_RANK
};

std::string trim(std::string_view value)
{
    std::size_t start = 0;
    while (start < value.size()
           && std::isspace(static_cast<unsigned char>(value[start])) != 0)
    {
        ++start;
    }

    std::size_t end = value.size();
    while (end > start && std::isspace(static_cast<unsigned char>(value[end - 1])) != 0)
    {
        --end;
    }

    return std::string(value.substr(start, end - start));
}

bool is_blank(std::string_view value)
{
    return trim(value).empty();
}

std::string normalize(std::string_view value)
{
    const std::string trimmed = trim(value);
    std::string result;
    result.reserve(trimmed.size());
    for (const unsigned char ch : trimmed)
    {
        if (ch >= 'A' && ch <= 'Z')
        {
            result.push_back(static_cast<char>(ch - 'A' + 'a'));
        }
        else
        {
            result.push_back(static_cast<char>(ch));
        }
    }
    return result;
}

bool contains(std::string_view value, std::string_view needle)
{
    return normalize(value).find(needle) != std::string::npos;
}

int popularity(std::string_view value)
{
    const std::string lower = normalize(value);
    for (const auto& rank : popularity_ranks)
        for (const auto keyword : rank.keywords)
            if (lower.find(keyword) != std::string::npos) return rank.score;
    return 0;
}

GameCenterState::Category category_from_name(std::string_view category_name)
{
    if (category_name == "RECENT")
    {
        return GameCenterState::Category::Recent;
    }
    if (category_name == "FAVORITES")
    {
        return GameCenterState::Category::Favorites;
    }
    if (category_name == "ALL")
    {
        return GameCenterState::Category::All;
    }
    if (category_name == "BUILTIN")
    {
        return GameCenterState::Category::Builtin;
    }
    return GameCenterState::Category::All;
}

} // namespace

int popularity_for_package(std::string_view outer_filename, std::string_view entry_path)
{
    const auto leaf = [](std::string_view value) {
        const auto separator = value.find_last_of("/\\");
        return separator == std::string_view::npos ? value : value.substr(separator + 1);
    };
    return std::max(popularity(leaf(outer_filename)), popularity(leaf(entry_path)));
}

GameCenterState GameCenterState::restore(const std::string& category_name,
                                         const std::string& query,
                                         const std::string& selected_canonical_id)
{
    GameCenterState state;
    state.category_ = category_from_name(category_name);
    state.query_ = query;
    if (!is_blank(selected_canonical_id))
    {
        state.selections_[state.category_] = selected_canonical_id;
    }
    return state;
}

std::string GameCenterState::selected_canonical_id() const
{
    const auto found = selections_.find(category_);
    if (found == selections_.end())
    {
        return {};
    }
    return found->second;
}

void GameCenterState::select(const std::string& canonical_id)
{
    if (is_blank(canonical_id))
    {
        selections_.erase(category_);
        return;
    }
    selections_[category_] = canonical_id;
}

std::vector<GameCenterItem> GameCenterState::items_for(
    Category value,
    const std::vector<GameCenterItem>& all) const
{
    std::vector<GameCenterItem> result;
    std::unordered_set<std::string> seen;
    for (const GameCenterItem& item : all)
    {
        bool include = false;
        switch (value)
        {
        case Category::Recent:
            include = item.last_played_sequence > 0;
            break;
        case Category::Favorites:
            include = item.favorite;
            break;
        case Category::All:
            include = true;
            break;
        case Category::Builtin:
            include = item.builtin;
            break;
        }
        // Native canonical IDs derive from exact ROM payload hashes, not game names.
        if (include && seen.insert(item.canonical_id).second)
        {
            result.push_back(item);
        }
    }

    if (value == Category::Recent)
    {
        std::sort(result.begin(),
                  result.end(),
                  [](const GameCenterItem& left, const GameCenterItem& right) {
                      return left.last_played_sequence > right.last_played_sequence;
                  });
    }
    else if (value == Category::All)
    {
        std::unordered_map<std::string, int> scores;
        for (const auto& item : all)
            scores[item.canonical_id] = std::max(scores[item.canonical_id],
                item.popularity_score >= 0 ? item.popularity_score
                    : std::max({popularity(item.title_en), popularity(item.title_zh_hans),
                                popularity(item.original_filename)}));
        std::sort(result.begin(), result.end(), [&](const GameCenterItem& left,
                                                    const GameCenterItem& right) {
            const int left_score = scores.at(left.canonical_id);
            const int right_score = scores.at(right.canonical_id);
            return left_score != right_score ? left_score > right_score
                                            : left.canonical_id < right.canonical_id;
        });
    }

    return result;
}

std::vector<GameCenterItem> GameCenterState::filtered(
    const std::vector<GameCenterItem>& all) const
{
    const std::vector<GameCenterItem> category_items = items_for(category_, all);
    const std::string needle = normalize(query_);
    if (needle.empty())
    {
        return category_items;
    }

    std::vector<GameCenterItem> result;
    for (const GameCenterItem& item : category_items)
    {
        if (contains(item.title_en, needle) || contains(item.title_zh_hans, needle)
            || contains(item.original_filename, needle))
        {
            result.push_back(item);
        }
    }
    return result;
}

void GameCenterState::reconcile(const std::vector<GameCenterItem>& visible)
{
    const std::string selected = selected_canonical_id();
    bool exists = false;
    for (const GameCenterItem& item : visible)
    {
        exists = exists || item.canonical_id == selected;
    }
    if (!exists)
    {
        select(visible.empty() ? std::string{} : visible.front().canonical_id);
    }
}

} // namespace flynes::product
